# Inspección inicial con Python
## Carga de librerías

In [3]:
import pandas as pd
import numpy as np
from pathlib import Path  # manejar rutas de archivos y carpetas de forma más cómoda

pd.set_option("display.max_columns", None)  # mostrar todas las columnas sin truncarlas con (...)

RAW_DIR = Path("../data/raw")  # variables que representan rutas
PROCESSED_DIR = Path("../data/processed")

## Carga de los archivos por separado

In [4]:
beds = pd.read_csv(RAW_DIR/"eurostat_hospital_beds.csv")
discharges = pd.read_csv(RAW_DIR/"eurostat_hospital_discharges.csv")
length_stay = pd.read_csv(RAW_DIR/"eurostat_average_length_stay.csv")

## Perfil de cada dataframe
### 1. Hospital beds

In [5]:
beds.shape

(30, 19)

In [6]:
beds.head()

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,freq,Time frequency,unit,Unit of measure,facility,Health facility,geo,Geopolitical entity (reporting),TIME_PERIOD,Time,OBS_VALUE,Observation value,OBS_FLAG,Observation status (Flag) V2 structure,CONF_STATUS,Confidentiality status (flag)
0,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),DE,Germany,2014,NaN,822.82,NaN,NaN,NaN,NaN,NaN
1,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),DE,Germany,2015,NaN,813.31,NaN,NaN,NaN,NaN,NaN
2,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),DE,Germany,2016,NaN,806.26,NaN,NaN,NaN,NaN,NaN
3,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),DE,Germany,2017,NaN,800.23,NaN,NaN,NaN,NaN,NaN
4,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),DE,Germany,2018,NaN,798.35,NaN,NaN,NaN,NaN,NaN


In [7]:
beds.sample(5)

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,freq,Time frequency,unit,Unit of measure,facility,Health facility,geo,Geopolitical entity (reporting),TIME_PERIOD,Time,OBS_VALUE,Observation value,OBS_FLAG,Observation status (Flag) V2 structure,CONF_STATUS,Confidentiality status (flag)
20,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),IT,Italy,2016,NaN,317.21,NaN,NaN,NaN,NaN,NaN
25,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),PT,Portugal,2015,NaN,336.84,NaN,e,estimated,NaN,NaN
19,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),IT,Italy,2015,NaN,319.55,NaN,NaN,NaN,NaN,NaN
14,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),FR,France,2016,NaN,605.88,NaN,NaN,NaN,NaN,NaN
0,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),DE,Germany,2014,NaN,822.82,NaN,NaN,NaN,NaN,NaN


In [8]:
beds.columns.tolist()

['STRUCTURE',
 'STRUCTURE_ID',
 'STRUCTURE_NAME',
 'freq',
 'Time frequency',
 'unit',
 'Unit of measure',
 'facility',
 'Health facility',
 'geo',
 'Geopolitical entity (reporting)',
 'TIME_PERIOD',
 'Time',
 'OBS_VALUE',
 'Observation value',
 'OBS_FLAG',
 'Observation status (Flag) V2 structure',
 'CONF_STATUS',
 'Confidentiality status (flag)']

In [9]:
def create_profile(df: pd.DataFrame) -> pd.DataFrame: 
    # def funcion(nombre_parametro: TipoEsperado) -> TipoQueDevuelve
    n_rows = len(df)
    cardinality = df.nunique(dropna=False)
    nulls = df.isna().sum()
        

    profile = pd.DataFrame({
        "dtype": df.dtypes.astype(str),
        "cardinality": cardinality,
        "unique_pct": (cardinality/n_rows*100).round(2),
        "nulls": df.isna().sum(),
        "null_pct": (df.isna().mean()*100).round(2),
        "duplicates": df.duplicated().sum(),
    })

    return profile

In [10]:
create_profile(beds)

,dtype,cardinality,unique_pct,nulls,null_pct,duplicates
STRUCTURE,str,1,3.33,0,0.00,0
STRUCTURE_ID,str,1,3.33,0,0.00,0
STRUCTURE_NAME,str,1,3.33,0,0.00,0
freq,str,1,3.33,0,0.00,0
Time frequency,str,1,3.33,0,0.00,0
unit,str,1,3.33,0,0.00,0
Unit of measure,str,1,3.33,0,0.00,0
facility,str,1,3.33,0,0.00,0
Health facility,str,1,3.33,0,0.00,0
geo,str,5,16.67,0,0.00,0


### 2. Hospital discharges

In [11]:
discharges.shape

(186, 25)

In [12]:
discharges.head()

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,freq,Time frequency,age,Age class,indic_he,Health indicator,unit,Unit of measure,sex,Sex,icd10,International Statistical Classification of Diseases and Related Health Problems (ICD-10),geo,Geopolitical entity (reporting),TIME_PERIOD,Time,OBS_VALUE,Observation value,OBS_FLAG,Observation status (Flag) V2 structure,CONF_STATUS,Confidentiality status (flag)
0,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2014,NaN,25602.5,NaN,NaN,NaN,NaN,NaN
1,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2015,NaN,25534.3,NaN,NaN,NaN,NaN,NaN
2,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2016,NaN,25685.6,NaN,NaN,NaN,NaN,NaN
3,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2017,NaN,25477.9,NaN,NaN,NaN,NaN,NaN
4,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2018,NaN,25246.8,NaN,NaN,NaN,NaN,NaN


In [13]:
discharges.sample(5)

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,freq,Time frequency,age,Age class,indic_he,Health indicator,unit,Unit of measure,sex,Sex,icd10,International Statistical Classification of Diseases and Related Health Problems (ICD-10),geo,Geopolitical entity (reporting),TIME_PERIOD,Time,OBS_VALUE,Observation value,OBS_FLAG,Observation status (Flag) V2 structure,CONF_STATUS,Confidentiality status (flag)
148,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,J,Diseases of the respiratory system (J00-J99),DE,Germany,2014,NaN,1528.2,NaN,NaN,NaN,NaN,NaN
170,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,J,Diseases of the respiratory system (J00-J99),FR,France,2020,NaN,780.6,NaN,NaN,NaN,NaN,NaN
173,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,J,Diseases of the respiratory system (J00-J99),IT,Italy,2015,NaN,1004.7,NaN,NaN,NaN,NaN,NaN
178,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,J,Diseases of the respiratory system (J00-J99),IT,Italy,2020,NaN,704.6,NaN,NaN,NaN,NaN,NaN
156,dataflow,ESTAT:HLTH_CO_DISCH2(1.0),"Hospital discharges by diagnosis, in-patients,...",A,Annual,TOTAL,Total,INPAT,in-patients (total number),P_HTHAB,Per hundred thousand inhabitants,T,Total,J,Diseases of the respiratory system (J00-J99),ES,Spain,2014,NaN,1142.7,NaN,NaN,NaN,NaN,NaN


In [14]:
discharges.columns.tolist()

['STRUCTURE',
 'STRUCTURE_ID',
 'STRUCTURE_NAME',
 'freq',
 'Time frequency',
 'age',
 'Age class',
 'indic_he',
 'Health indicator',
 'unit',
 'Unit of measure',
 'sex',
 'Sex',
 'icd10',
 'International Statistical Classification of Diseases and Related Health Problems (ICD-10)',
 'geo',
 'Geopolitical entity (reporting)',
 'TIME_PERIOD',
 'Time',
 'OBS_VALUE',
 'Observation value',
 'OBS_FLAG',
 'Observation status (Flag) V2 structure',
 'CONF_STATUS',
 'Confidentiality status (flag)']

In [15]:
create_profile(discharges)

,dtype,cardinality,unique_pct,nulls,null_pct,duplicates
STRUCTURE,str,1,0.54,0,0.0,0
STRUCTURE_ID,str,1,0.54,0,0.0,0
STRUCTURE_NAME,str,1,0.54,0,0.0,0
freq,str,1,0.54,0,0.0,0
Time frequency,str,1,0.54,0,0.0,0
age,str,1,0.54,0,0.0,0
Age class,str,1,0.54,0,0.0,0
indic_he,str,1,0.54,0,0.0,0
Health indicator,str,1,0.54,0,0.0,0
unit,str,1,0.54,0,0.0,0


### 3. Average lenght of stay

In [16]:
length_stay.shape

(188, 25)

In [17]:
length_stay.head()

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,freq,Time frequency,age,Age class,indic_he,Health indicator,unit,Unit of measure,sex,Sex,icd10,International Statistical Classification of Diseases and Related Health Problems (ICD-10),geo,Geopolitical entity (reporting),TIME_PERIOD,Time,OBS_VALUE,Observation value,OBS_FLAG,Observation status (Flag) V2 structure,CONF_STATUS,Confidentiality status (flag)
0,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2014,NaN,9.0,NaN,NaN,NaN,NaN,NaN
1,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2015,NaN,9.0,NaN,NaN,NaN,NaN,NaN
2,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2016,NaN,8.9,NaN,NaN,NaN,NaN,NaN
3,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2017,NaN,8.9,NaN,NaN,NaN,NaN,NaN
4,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,DE,Germany,2018,NaN,8.9,NaN,NaN,NaN,NaN,NaN


In [18]:
length_stay.sample(5)

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,freq,Time frequency,age,Age class,indic_he,Health indicator,unit,Unit of measure,sex,Sex,icd10,International Statistical Classification of Diseases and Related Health Problems (ICD-10),geo,Geopolitical entity (reporting),TIME_PERIOD,Time,OBS_VALUE,Observation value,OBS_FLAG,Observation status (Flag) V2 structure,CONF_STATUS,Confidentiality status (flag)
40,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,C00-D48,Neoplasms,DE,Germany,2018,NaN,8.9,NaN,NaN,NaN,NaN,NaN
168,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,J,Diseases of the respiratory system (J00-J99),FR,France,2016,NaN,6.7,NaN,NaN,NaN,NaN,NaN
19,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,A-T_Z,All causes of diseases (A00-Z99) excluding V00...,FR,France,2018,NaN,8.8,NaN,NaN,NaN,NaN,NaN
98,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,F,Mental and behavioural disorders (F00-F99),IT,Italy,2014,NaN,13.9,NaN,NaN,NaN,NaN,NaN
45,dataflow,ESTAT:HLTH_CO_INPST(1.0),In-patient average length of stay (days)- hist...,A,Annual,TOTAL,Total,ALOS,In-patient average length of stay (in days),NR,Number,T,Total,C00-D48,Neoplasms,ES,Spain,2015,NaN,8.2,NaN,NaN,NaN,NaN,NaN


In [19]:
length_stay.columns.tolist()

['STRUCTURE',
 'STRUCTURE_ID',
 'STRUCTURE_NAME',
 'freq',
 'Time frequency',
 'age',
 'Age class',
 'indic_he',
 'Health indicator',
 'unit',
 'Unit of measure',
 'sex',
 'Sex',
 'icd10',
 'International Statistical Classification of Diseases and Related Health Problems (ICD-10)',
 'geo',
 'Geopolitical entity (reporting)',
 'TIME_PERIOD',
 'Time',
 'OBS_VALUE',
 'Observation value',
 'OBS_FLAG',
 'Observation status (Flag) V2 structure',
 'CONF_STATUS',
 'Confidentiality status (flag)']

In [20]:
create_profile(length_stay)

,dtype,cardinality,unique_pct,nulls,null_pct,duplicates
STRUCTURE,str,1,0.53,0,0.0,0
STRUCTURE_ID,str,1,0.53,0,0.0,0
STRUCTURE_NAME,str,1,0.53,0,0.0,0
freq,str,1,0.53,0,0.0,0
Time frequency,str,1,0.53,0,0.0,0
age,str,1,0.53,0,0.0,0
Age class,str,1,0.53,0,0.0,0
indic_he,str,1,0.53,0,0.0,0
Health indicator,str,1,0.53,0,0.0,0
unit,str,1,0.53,0,0.0,0


## Tablas descriptivas
### Beds

| Column | Description | Variable type | Cardinality | Missing values | Importance | Notes |
|-|-|-|-|-|-|-|
| `STRUCTURE` | Eurostat statistical structure family | Technical categorical | 1 | 0 (0%) | Low | Source metadata; drop after profiling. |
| `STRUCTURE_ID` | Technical identifier of the statistical structure | Technical identifier | 1 | 0 (0%) | Low | Drop. |
| `STRUCTURE_NAME` | Human-readable name of the statistical structure | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `STRUCTURE_ID`; drop. |
| `freq` | Time-frequency code | Categorical | 1 | 0 (0%) | Medium | Confirms annual frequency; drop. |
| `Time frequency` | Human-readable time-frequency label | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `freq`; drop. |
| `unit` | Unit-of-measure code | Categorical | 1 | 0 (0%) | High | Must represent beds per 100,000 inhabitants. Retain initially for traceability. |
| `Unit of measure` | Human-readable unit label | Descriptive text | 1 | 0 (0%) | Medium | Useful for documentation; drop from the final fact table if the unit is already documented. |
| `facility` | Health-facility type code | Categorical | 1 | 0 (0%) | High | Represents the selected filter: hospitals. Document it; as it is constant, it may be dropped from the final fact table. |
| `Health facility` | Human-readable health-facility type | Descriptive text | 1 | 0 (0%) | Medium | Must correspond to `Hospitals`; descriptive duplicate of `facility`. |
| `geo` | Reporting-country code | Categorical / dimension key | 5 | 0 (0%) | Critical | ES, PT, FR, IT and DE. It will be mapped to `country_id` through `dim_country`. |
| `Geopolitical entity (reporting)` | Reporting-country name | Descriptive text | 5 | 0 (0%) | Medium | Useful for validation. In the final model, it will come from `dim_country`. |
| `TIME_PERIOD` | Observation year | Discrete temporal / dimension key | 6 | 0 (0%) | Critical | Verify the six years included and whether coverage is complete for all five countries. |
| `Time` | Auxiliary date field | Empty numeric field | 1 | 30 (100%) | Not applicable | Empty in the export; drop. |
| `OBS_VALUE` | Hospital beds per 100,000 inhabitants | Continuous numeric | 30 | 0 (0%) | Critical | Main metric; do not sum across countries or years. |
| `Observation value` | Auxiliary descriptive value field | Empty numeric field | 1 | 30 (100%) | Not applicable | Empty in the export; drop. |
| `OBS_FLAG` | Observation-status flag | Categorical | 2 | 29 (96.67%) | High | Retain as `status_flag`. A null value generally means that no flag was reported. |
| `Observation status (Flag) V2 structure` | Human-readable observation-status flag | Descriptive text | 2 | 29 (96.67%) | Low | Descriptive duplicate of `OBS_FLAG`; drop. |
| `CONF_STATUS` | Confidentiality-status flag | Empty numeric field | 1 | 30 (100%) | Not applicable | No confidentiality information is reported; drop. |
| `Confidentiality status (flag)` | Human-readable confidentiality-status flag | Empty numeric field | 1 | 30 (100%) | Not applicable | Empty in the export; drop. |


### Discharges

| Column | Description | Variable type | Cardinality | Missing values | Importance | Notes |
|-|-|-|-|-|-|-|
| `STRUCTURE` | Eurostat statistical structure family | Technical categorical | 1 | 0 (0%) | Low | Source metadata; document and drop after profiling. |
| `STRUCTURE_ID` | Technical identifier of the statistical structure | Technical identifier | 1 | 0 (0%) | Low | Document and drop after profiling. |
| `STRUCTURE_NAME` | Human-readable name of the statistical structure | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `STRUCTURE_ID`; drop. |
| `freq` | Time-frequency code | Categorical | 1 | 0 (0%) | Medium | Confirms annual frequency; drop. |
| `Time frequency` | Human-readable time-frequency label | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `freq`; drop. |
| `age` | Age-group code | Categorical | 1 | 0 (0%) | Medium | Represents the total-age filter. Document it; as it is constant, it may be dropped from the final fact table. |
| `Age class` | Human-readable age-group label | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `age`; drop. |
| `indic_he` | Health-indicator code | Categorical | 1 | 0 (0%) | High | Must correspond to in-patient hospital discharges per 100,000 inhabitants. Document it as part of the metric definition. |
| `Health indicator` | Human-readable health-indicator label | Descriptive text | 1 | 0 (0%) | Medium | Descriptive duplicate of `indic_he`; drop after documenting the filter. |
| `unit` | Unit-of-measure code | Categorical | 1 | 0 (0%) | High | Must represent discharges per 100,000 inhabitants. Retain initially for traceability. |
| `Unit of measure` | Human-readable unit label | Descriptive text | 1 | 0 (0%) | Medium | Useful for documentation; drop from the final fact table if the unit is already defined in the metric. |
| `sex` | Sex code | Categorical | 1 | 0 (0%) | Medium | Represents the total-sex filter. Document it; as it is constant, it may be dropped. |
| `Sex` | Human-readable sex label | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `sex`; drop. |
| `icd10` | ICD-10 diagnosis-group code | Categorical / dimension key | 5 | 0 (0%) | Critical | All diseases plus neoplasms, mental and behavioural disorders, circulatory diseases and respiratory diseases. Basis for `dim_diagnosis`. Do not sum diagnoses to reconstruct the total. |
| `International Statistical Classification of Diseases and Related Health Problems (ICD-10)` | Diagnosis-group name | Descriptive text | 5 | 0 (0%) | High | Use it to create the label in `dim_diagnosis`; do not retain the duplicate in the fact table. |
| `geo` | Reporting-country code | Categorical / dimension key | 5 | 0 (0%) | Critical | ES, PT, FR, IT and DE. It will be mapped to `country_id` through `dim_country`. |
| `Geopolitical entity (reporting)` | Reporting-country name | Descriptive text | 5 | 0 (0%) | Medium | Useful for validation. In the final model, it will come from `dim_country`. |
| `TIME_PERIOD` | Observation year | Discrete temporal / dimension key | 8 | 0 (0%) | Critical | Verify the exact years and coverage by country and diagnosis. |
| `Time` | Auxiliary date field | Empty numeric field | 1 | 186 (100%) | Not applicable | Empty in the export; drop. |
| `OBS_VALUE` | In-patient hospital discharges per 100,000 inhabitants | Continuous numeric | 186 | 0 (0%) | Critical | Main metric. Do not sum across countries, years or diagnoses. |
| `Observation value` | Auxiliary descriptive value field | Empty numeric field | 1 | 186 (100%) | Not applicable | Empty in the export; drop. |
| `OBS_FLAG` | Observation-status flag | Empty numeric field | 1 | 186 (100%) | Medium | No observation-status flags are reported in this extract. Document and drop. |
| `Observation status (Flag) V2 structure` | Human-readable observation-status flag | Empty numeric field | 1 | 186 (100%) | Not applicable | Empty in the export; drop. |
| `CONF_STATUS` | Confidentiality-status flag | Empty numeric field | 1 | 186 (100%) | Not applicable | No confidentiality information is reported. Document and drop. |
| `Confidentiality status (flag)` | Human-readable confidentiality-status flag | Empty numeric field | 1 | 186 (100%) | Not applicable | Empty in the export; drop. |

### Average length of stay

| Column | Description | Variable type | Cardinality | Missing values | Importance | Notes |
|-|-|-|-|-|-|-|
| `STRUCTURE` | Eurostat statistical structure family | Technical categorical | 1 | 0 (0%) | Low | Source metadata; document and drop after profiling. |
| `STRUCTURE_ID` | Technical identifier of the statistical structure | Technical identifier | 1 | 0 (0%) | Low | Document and drop after profiling. |
| `STRUCTURE_NAME` | Human-readable name of the statistical structure | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `STRUCTURE_ID`; drop. |
| `freq` | Time-frequency code | Categorical | 1 | 0 (0%) | Medium | Confirms annual frequency; document and drop from the processed dataset. |
| `Time frequency` | Human-readable time-frequency label | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `freq`; drop. |
| `age` | Age-group code | Categorical | 1 | 0 (0%) | Medium | Represents the total-age filter. Document it; as it is constant, it may be dropped. |
| `Age class` | Human-readable age-group label | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `age`; drop. |
| `indic_he` | Health-indicator code | Categorical | 1 | 0 (0%) | High | Identifies the average length-of-stay indicator; document it as part of the metric definition. |
| `Health indicator` | Human-readable health-indicator label | Descriptive text | 1 | 0 (0%) | Medium | Descriptive duplicate of `indic_he`; drop after documenting the filter. |
| `unit` | Unit-of-measure code | Categorical | 1 | 0 (0%) | High | Eurostat may label it as `Number`; it will be interpreted and presented as average days per discharge. |
| `Unit of measure` | Human-readable unit label | Descriptive text | 1 | 0 (0%) | High | Document that this is not a rate per 100,000 inhabitants and must not be summed. |
| `sex` | Sex code | Categorical | 1 | 0 (0%) | Medium | Represents the total-sex filter. Document it; as it is constant, it may be dropped. |
| `Sex` | Human-readable sex label | Descriptive text | 1 | 0 (0%) | Low | Descriptive duplicate of `sex`; drop. |
| `icd10` | ICD-10 diagnosis-group code | Categorical / dimension key | 5 | 0 (0%) | Critical | All diseases, neoplasms, mental and behavioural disorders, circulatory diseases and respiratory diseases. Basis for `dim_diagnosis`. |
| `International Statistical Classification of Diseases and Related Health Problems (ICD-10)` | Diagnosis-group name | Descriptive text | 5 | 0 (0%) | High | Use it to create `dim_diagnosis`; do not retain the duplicate in the fact table. |
| `geo` | Reporting-country code | Categorical / dimension key | 5 | 0 (0%) | Critical | DE, ES, FR, IT and PT. It will be mapped to `country_id` through `dim_country`. |
| `Geopolitical entity (reporting)` | Reporting-country name | Descriptive text | 5 | 0 (0%) | Medium | Useful for validation. In the final model, it will come from `dim_country`. |
| `TIME_PERIOD` | Observation year | Discrete temporal / dimension key | 8 | 0 (0%) | Critical | Coverage spans 2014–2021; the project will initially use 2014–2019 to align with the beds dataset. |
| `Time` | Auxiliary date field | Empty numeric field | 1 | 190 (100%) | Not applicable | Empty in the export; drop. |
| `OBS_VALUE` | Average length of stay for in-patients | Continuous numeric | 73 | 0 (0%) | Critical | Average value in days per discharge. Do not sum across countries, years or diagnoses. |
| `Observation value` | Auxiliary descriptive value field | Empty numeric field | 1 | 190 (100%) | Not applicable | Empty in the export; drop. |
| `OBS_FLAG` | Observation-status flag | Empty numeric field | 1 | 190 (100%) | Medium | No observation-status flags are reported in this extract. Document and drop. |
| `Observation status (Flag) V2 structure` | Human-readable observation-status flag | Empty numeric field | 1 | 190 (100%) | Not applicable | Empty in the export; drop. |
| `CONF_STATUS` | Confidentiality-status flag | Empty numeric field | 1 | 190 (100%) | Not applicable | No confidentiality information is reported. Document and drop. |
| `Confidentiality status (flag)` | Human-readable confidentiality-status flag | Empty numeric field | 1 | 190 (100%) | Not applicable | Empty in the export; drop. |

## Data coverage assesment

The coverage tables show whether a valid observation exists for each expected country-year combination and, for activity indicators, each country-year-diagnosis combination. A value of `1` indicates an available observation; `0` indicates that Eurostat does not provide an observation for that combiantion.


In [25]:
beds.sample(5)

,STRUCTURE,STRUCTURE_ID,STRUCTURE_NAME,freq,Time frequency,unit,Unit of measure,facility,Health facility,geo,Geopolitical entity (reporting),TIME_PERIOD,Time,OBS_VALUE,Observation value,OBS_FLAG,Observation status (Flag) V2 structure,CONF_STATUS,Confidentiality status (flag)
17,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),FR,France,2019,NaN,583.79,NaN,NaN,NaN,NaN,NaN
19,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),IT,Italy,2015,NaN,319.55,NaN,NaN,NaN,NaN,NaN
23,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),IT,Italy,2019,NaN,316.28,NaN,NaN,NaN,NaN,NaN
10,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),ES,Spain,2018,NaN,297.15,NaN,NaN,NaN,NaN,NaN
15,dataflow,ESTAT:HLTH_RS_BDS(1.0),Hospital beds by type of care - historical dat...,A,Annual,P_HTHAB,Per hundred thousand inhabitants,HBEDT,Available beds in hospitals (HP.1),FR,France,2017,NaN,598.02,NaN,NaN,NaN,NaN,NaN


In [22]:
beds_coverage = (
    beds
    .pivot_table(
        index="geo",
        columns="TIME_PERIOD",
        values="OBS_VALUE",
        aggfunc="count"
    )
    .fillna(0)  
    .astype(int)  # method chaining
)

beds_coverage

TIME_PERIOD,2014,2015,2016,2017,2018,2019
geo,,,,,,
DE,1,1,1,1,1,1
ES,1,1,1,1,1,1
FR,1,1,1,1,1,1
IT,1,1,1,1,1,1
PT,1,1,1,1,1,1


In [23]:
discharges_coverage = (
    discharges
    .pivot_table(
        index=["geo", "icd10"],
        columns="TIME_PERIOD",
        values="OBS_VALUE",
        aggfunc="count"
    )
    .fillna(0)
    .astype(int)
)

discharges_coverage

TIME_PERIOD  2014  2015  2016  2017  2018  2019  2020  2021
geo icd10                                                  
DE  A-T_Z       1     1     1     1     1     1     1     1
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
ES  A-T_Z       0     1     1     1     1     1     1     1
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
FR  A-T_Z       1     1     1     1     1     1     1     1
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
IT  A-T_Z       1     1     1     1     1     0     0     0
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
PT  A-T_Z       1     1     0     0     1     1     1     1
    C00-D48     1     1     0     0     1     1     1     1
    F           1     1     0     0     1     1     1     1
    I           1     1     0     0     1     1     1     1
    J           1     1     0     0     1     1     1     1

In [24]:
length_stay_coverage = (
    length_stay
    .pivot_table(
        index=["geo", "icd10"],
        columns="TIME_PERIOD",
        values="OBS_VALUE",
        aggfunc="count"
    )
    .fillna(0)
    .astype(int)
)

length_stay_coverage

TIME_PERIOD  2014  2015  2016  2017  2018  2019  2020  2021
geo icd10                                                  
DE  A-T_Z       1     1     1     1     1     1     1     1
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
ES  A-T_Z       0     1     1     1     1     1     1     1
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
FR  A-T_Z       1     1     1     1     1     1     1     1
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
IT  A-T_Z       1     1     1     1     1     1     1     0
    C00-D48     1     1     1     1     1     1     1     1
    F           1     1     1     1     1     1     1     1
    I           1     1     1     1     1     1     1     1
    J           1     1     1     1     1     1     1     1
PT  A-T_Z       1     1     0     0     1     1     1     1
    C00-D48     1     1     0     0     1     1     1     1
    F           1     1     0     0     1     1     1     1
    I           1     1     0     0     1     1     1     1
    J           1     1     0     0     1     1     1     1

- Hospital beds have complete coverage for Germany, Spain, France, Italy and Portugal from 2014 to 2019.
- Hospital discharges and average length of stay: cover a wider historical period (2014-21) bur their availability is not fully balanced across all countries and diagnoses. Spain lacks the all-diseases observation in 2014 and Italy from 2019 onwards. Portugal has no observations for any diagnosis in 2016-17.
- The **project analysis period** is therefore set to **2014-19**, which aligns with the complete historical hospital-beds series. Missing observations will be retained as absent data and documented; they will not be imputed.


## Profiling conclusions  

### 1. What does each row represent?  

- `beds`: one observation per reporting country and year, representing total hospital beds in hospitals per 100,000 inhabitants.
- `discharges`: one observation per reporting country, year and ICD-10 diagnosis group, representing in-patient hospital discharges per 100,000 inhabitants.
- `lenght_stay`: one observation per reporting country, year and ICD-10 diagnosis group, representing the average lenght of stay for in-patients in days. 

### 2. Which columns uniquely identify an observation?

The raw datasets contain several metadata columns that are constant after filtering. The analytical keys are: 

- `beds`: `geo`+ `TIME_PERIOD`
- `discharges`: `geo`+ `TIME_PERIOD` + `icd-10`
- `lenght_stay`: `geo`+ `TIME_PERIOD` + `icd-10`

These combinations must be unique after cleaning and will define the granularity of the fact tables. 

### 3. Are years stored in rows or columns?

Years are stored in rows through the `TIME_PERIOD`column. The coverage tables pivot years into columns only to make missing country-year combinations easier to inspect. 

### 4. How are missing values represented?

There are no null values in `OBS_VALUE`in the downloaded records. However, some expected country-year-diagnosis combinations are absent entirely from the source data. These structural absences are identified through the coverage tables and must not be treated as zero values. 

### 5. Is there a separate observation-status flag?

Yes. Eurostat provides `OBS_FLAG`and a corresponding descriptive label column. Only in `beds`, one observation has a status flag. 

The raw `OBS_FLAG` column will be retained initially as `status_flag` where it contains information. Empty flag columns will be removed from the final processed datasets. 

### 6. Are the rates already calculated per 100,000 inhabitants?

- Hospital beds and hospital discharges are already reported per 100,000
  inhabitants.
- Average length of stay is not a population rate. It is an average duration in
  days per hospital discharge.

Therefore, beds and discharges must not be summed across countries or years,
and average length of stay must not be summed across diagnoses, countries or
years.

### 7. Are there multiple units within a dataset?

No. Each downloaded dataset has one unit after filtering.

- Beds: per 100,000 inhabitants.
- Discharges: per 100,000 inhabitants.
- Average length of stay: days.

### 8. What diagnosis codes are included?

The selected diagnosis groups are:

- `A-T_Z`: All diseases
- `C00-D48`: Neoplasms
- `F`: Mental and behavioural disorders
- `I`: Diseases of the circulatory system
- `J`: Diseases of the respiratory system

### 9. Do totals and subcategories create a risk of double counting?

Yes. `A-T_Z` represents all diseases, while `C00-D48`, `F`, `I` and `J`
represent diagnostic subgroups included within the total.

The total must never be reconstructed by summing the selected diagnoses, and
total observations must not be added to subgroup observations.

### 11. What is the final time scope?

The latest year shared by the three source datasets is 2019 because the
hospital-beds historical dataset ends in that year.

The project will analyse 2014–2019. Comparisons involving Portugal in 2016 and
2017, and all-diseases discharge comparisons where coverage is incomplete,
will be excluded or explicitly labelled as incomplete.